Source/WebKit/WebProcess/Plugins/PluginView.cpp

```diff
@@ -534,7 +532,28 @@ void PluginView::paint(GraphicsContext& context, const IntRect& dirtyRect, Widge
         return;
     }
 
-    m_plugin->paint(context, dirtyRect);
+    bool isSnapshotting = [&]() {
+        auto* frameView = frame()->view();
+        if (!frameView)
+            return false;
+
+        return frameView->paintBehavior().contains(PaintBehavior::FlattenCompositingLayers);
+    }();
+
+    if (!isSnapshotting && m_plugin->layerHostingStrategy() == PluginLayerHostingStrategy::GraphicsLayer)
+        return;
+
+    // RenderWidget::paintContents() translated the context so the origin is aligned with frameRect().location().
+    // Shift it back to align with the plugin origin.
+    GraphicsContextStateSaver stateSaver(context);
+
+    auto widgetOrigin = frameRect().location();
+    context.translate(widgetOrigin);
+
+    auto paintRect = dirtyRect;
+    paintRect.moveBy(-widgetOrigin);
+
+    m_plugin->paint(context, paintRect);
```

Comment:
We should have a RefPtr to the plugin on the stack before we call paint() on it. something like:
```
protectedPlugin()->paint(...);
```